File: benchmarks/ackermann.fs
\ Ackermann(3,10) benchmark - deep recursion
: ackermann ( m n -- result )
  over 0= if nip 1+ exit then
  dup 0= if drop 1- 1 recurse exit then
  over swap 1- recurse   \ m ack(m, n-1)
  swap 1- swap recurse ;

: main
  3 10 ackermann . cr ;

File: benchmarks/ackermann.c
// Ackermann(3,10) benchmark - deep recursion
#include <stdio.h>

long ackermann(long m, long n) {
    if (m == 0) return n + 1;
    if (n == 0) return ackermann(m - 1, 1);
    return ackermann(m - 1, ackermann(m, n - 1));
}

int main() {
    long result = ackermann(3, 10);
    printf("%ld\n", result);
    return 0;
}
